src/permission.ts

```diff
@@ -36,133 +37,123 @@ export interface ActionData {
     authorized_by: NameType
 }
 
-export class Permission {
-    permission_name: Name
-    permission_data: PermissionData
+export type PermissionType =
+    | Permission
+    | API.v1.AccountPermission
+    | {
+          perm_name: NameType
+          parent: NameType
+          required_auth: AuthorityType
+          linked_actions?: LinkedActionType[]
+      }
 
-    constructor(permissionName: NameType, permissionData: PermissionData) {
-        this.permission_name = Name.from(permissionName)
-        this.permission_data = permissionData
-    }
+export type LinkedActionType = LinkedAction | {account: NameType; action: NameType}
 
-    get permissionName(): Name {
-        return this.permission_name
-    }
+@Struct.type('linked_actions')
+export class LinkedAction extends Struct {
+    @Struct.field('name') declare account: Name
+    @Struct.field('name') declare action: Name
+}
+
+export type WaitWeightType = WaitWeight | {wait_sec: UInt32Type; weight: UInt16Type}
 
-    get actionData(): ActionData {
-        return {
-            ...this.permission_data,
-            auth: Authority.from({
-                keys: this.permission_data.auth?.keys?.map(({key, weight}) => {
-                    return {
-                        key: String(key),
-                        weight: Number(weight),
-                    }
-                }),
-                accounts: this.permission_data.auth?.accounts?.map(({permission, weight}) => ({
-                    permission: {
-                        actor: String(permission.actor),
-                        permission: String(permission.permission),
-                    },
-                    weight: Number(weight),
-                })),
-                waits: this.permission_data.auth?.waits?.map(({wait_sec, weight}) => ({
-                    wait_sec: Number(wait_sec),
-                    weight: Number(weight),
-                })),
-                threshold: Number(this.permission_data.auth?.threshold),
-            }),
+@Struct.type('permission')
+export class Permission extends Struct {
```

Comment:
This is so much better! Not sure why I didn't think of making Permission a Struct 😅